preview/MsixCore/MsixCoreInstaller/FirewallRules.cpp

```diff
@@ -0,0 +1,254 @@
+
+HRESULT FirewallRules::ExecuteForAddRequest()
+{
+    for (auto firewallRule = m_firewallRules.begin(); firewallRule != m_firewallRules.end(); ++firewallRule)
+    {
+        RETURN_IF_FAILED(AddFirewallRules(*firewallRule));
+    }
+    return S_OK;
+}
+
+HRESULT FirewallRules::ParseManifest()
+{
+    ComPtr<IMsixDocumentElement> domElement;
+    RETURN_IF_FAILED(m_msixRequest->GetPackageInfo()->GetManifestReader()->QueryInterface(UuidOfImpl<IMsixDocumentElement>::iid, reinterpret_cast<void**>(&domElement)));
+
+    ComPtr<IMsixElement> element;
+    RETURN_IF_FAILED(domElement->GetDocumentElement(&element));
+
+    ComPtr<IMsixElementEnumerator> firewallRuleEnum;
+    RETURN_IF_FAILED(element->GetElements(firewallExtensionQuery.c_str(), &firewallRuleEnum));
+    BOOL hasCurrent = FALSE;
+    RETURN_IF_FAILED(firewallRuleEnum->GetHasCurrent(&hasCurrent));
+    while (hasCurrent)
+    {
+        FirewallRule firewallRule;
+        ComPtr<IMsixElement> ruleElement;
+        RETURN_IF_FAILED(firewallRuleEnum->GetCurrent(&ruleElement));
+
+        Text<wchar_t> direction;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(firewallRuleDirectionAttribute.c_str(), &direction));
+        firewallRule.direction = direction.Get();
+
+        Text<wchar_t> protocol;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(protocolAttribute.c_str(), &protocol));
+        firewallRule.protocol = protocol.Get();
+
+        Text<wchar_t> profile;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(profileAttribute.c_str(), &profile));
+        firewallRule.profile = profile.Get();
+
+        Text<wchar_t> localPortMin, localPortMax, remotePortMin, remotePortMax;
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(localPortMinAttribute.c_str(), &localPortMin));
+        if (localPortMin.Get() != nullptr)
+        {
+            firewallRule.localPortMin = localPortMin.Get();
+        }
+
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(localPortMaxAttribute.c_str(), &localPortMax));
+        if (localPortMax.Get() != nullptr)
+        {
+            firewallRule.localPortMax = localPortMax.Get();
+        }
+
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(remotePortMinAttribute.c_str(), &remotePortMin));
+        if (remotePortMin.Get() != nullptr)
+        {
+            firewallRule.remotePortMin = remotePortMin.Get();
+        }
+
+        RETURN_IF_FAILED(ruleElement->GetAttributeValue(remotePortMaxAttribute.c_str(), &remotePortMax));
+        if (remotePortMax.Get() != nullptr)
+        {
+            firewallRule.remotePortMax = remotePortMax.Get();
+        }
+
+        m_firewallRules.push_back(firewallRule);
+
+        RETURN_IF_FAILED(firewallRuleEnum->MoveNext(&hasCurrent));
+    }
+
+    return S_OK;
+}
+
+HRESULT FirewallRules::AddFirewallRules(FirewallRule& firewallRule)
+{
+    INetFwPolicy2 *pNetFwPolicy2 = NULL;
+    INetFwRules *pFwRules = NULL;
+    INetFwRule *pFwRule = NULL;

```

Comment:
these should be wrapped in com pointers to release them when out of scope

Reply:
Updated to use com pointers.